// @ngneat/query/esm2022/lib/infinite-query.mjs
import { assertInInjectionContext, inject, Injectable, Injector, runInInjectionContext, } from '@angular/core';
import { injectQueryClient } from './query-client';
import { InfiniteQueryObserver, } from '@tanstack/query-core';
import { createBaseQuery, } from './base-query';
import * as i0 from "@angular/core";
class InfiniteQuery {
    #instance = injectQueryClient();
    #injector = inject(Injector);
    use(options) {
        return createBaseQuery({
            client: this.#instance,
            injector: options.injector ?? this.#injector,
            Observer: InfiniteQueryObserver,
            options: options,
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.1.3", ngImport: i0, type: InfiniteQuery, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.1.3", ngImport: i0, type: InfiniteQuery, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.1.3", ngImport: i0, type: InfiniteQuery, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }] });
export function injectInfiniteQuery(options) {
    if (options?.injector) {
        return runInInjectionContext(options.injector, () => {
            const query = inject(InfiniteQuery);
            return query.use.bind(query);
        });
    }
    assertInInjectionContext(injectInfiniteQuery);
    const query = inject(InfiniteQuery);
    return query.use.bind(query);
}
