// @ngneat/query/esm2022/lib/is-fetching.mjs
import { notifyManager } from '@tanstack/query-core';
import { injectQueryClient } from './query-client';
import { assertInInjectionContext, inject, Injectable, InjectionToken, } from '@angular/core';
import { distinctUntilChanged, Observable } from 'rxjs';
import { toSignal } from '@angular/core/rxjs-interop';
import * as i0 from "@angular/core";
export class IsFetching {
    #queryClient = injectQueryClient();
    use(filters) {
        const result$ = new Observable((observer) => {
            observer.next(this.#queryClient.isFetching(filters));
            const disposeSubscription = this.#queryClient.getQueryCache().subscribe(notifyManager.batchCalls(() => {
                observer.next(this.#queryClient.isFetching(filters));
            }));
            return () => disposeSubscription();
        }).pipe(distinctUntilChanged());
        return {
            result$,
            toSignal: () => toSignal(result$),
        };
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.1.3", ngImport: i0, type: IsFetching, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.1.3", ngImport: i0, type: IsFetching, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.1.3", ngImport: i0, type: IsFetching, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }] });
const UseIsFetching = new InjectionToken('UseIsFetching', {
    providedIn: 'root',
    factory() {
        const isFetching = new IsFetching();
        return isFetching.use.bind(isFetching);
    },
});
export function injectIsFetching() {
    assertInInjectionContext(injectIsFetching);
    return inject(UseIsFetching);
}
