// @ngneat/query/esm2022/lib/mutation.mjs
import { inject, Injectable, InjectionToken } from '@angular/core';
import { injectQueryClient } from './query-client';
import { MutationObserver, notifyManager, } from '@tanstack/query-core';
import { isObservable, Observable, shareReplay } from 'rxjs';
import { toSignal } from '@angular/core/rxjs-interop';
import { shouldThrowError, toPromise } from './utils';
import * as i0 from "@angular/core";
class Mutation {
    #instance = injectQueryClient();
    use(options) {
        const mutationObserver = new MutationObserver(this.#instance, {
            ...options,
            mutationFn: (variables) => {
                const source = options.mutationFn(variables);
                if (isObservable(source))
                    return toPromise({ source });
                return source;
            },
        });
        const result$ = new Observable((observer) => {
            observer.next(mutationObserver.getCurrentResult());
            const disposeSubscription = mutationObserver.subscribe(notifyManager.batchCalls((result) => {
                if (result.isError &&
                    shouldThrowError(mutationObserver.options.throwOnError, [
                        result.error,
                    ])) {
                    observer.error(result.error);
                }
                else {
                    observer.next(result);
                }
            }));
            return () => disposeSubscription();
        }).pipe(shareReplay({
            bufferSize: 1,
            refCount: true,
        }));
        const mutate = (variables, options) => {
            mutationObserver.mutate(variables, options).catch(() => {
                // noop
            });
        };
        let cachedSignal;
        return {
            mutate,
            mutateAsync: mutationObserver.mutate.bind(mutationObserver),
            reset: mutationObserver.reset.bind(mutationObserver),
            setOptions: mutationObserver.setOptions.bind(mutationObserver),
            result$,
            // @experimental signal support
            get result() {
                if (!cachedSignal) {
                    cachedSignal = toSignal(this.result$, { requireSync: true });
                }
                return cachedSignal;
            },
        };
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.1.3", ngImport: i0, type: Mutation, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.1.3", ngImport: i0, type: Mutation, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.1.3", ngImport: i0, type: Mutation, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }] });
const UseMutation = new InjectionToken('UseMutation', {
    providedIn: 'root',
    factory() {
        const mutation = new Mutation();
        return mutation.use.bind(mutation);
    },
});
export function injectMutation() {
    return inject(UseMutation);
}
