// @ngneat/query/esm2022/lib/operators.mjs
import { filter, map, startWith, takeWhile, tap, } from 'rxjs';
import { createPendingObserverResult } from './utils';
export function mapResultData(mapFn) {
    return map((result) => {
        return {
            ...result,
            data: result.isSuccess
                ? mapFn(result.data)
                : result.data,
        };
    });
}
export function filterSuccessResult() {
    return filter((result) => result.isSuccess);
}
export function filterErrorResult() {
    return filter((result) => result.status === 'error');
}
export function tapSuccessResult(cb) {
    return tap((result) => {
        if (result.isSuccess) {
            cb(result.data);
        }
    });
}
export function tapErrorResult(cb) {
    return tap((result) => {
        if (result.isError) {
            cb(result.error);
        }
    });
}
/**
 * An operator that takes values emitted by the source observable
 * until the `isFetching` property on the result is false.
 * It is intended to be used in scenarios where an observable stream should be listened to
 * until the result has finished fetching (e.g success or error).
 */
export function takeUntilResultFinalize() {
    return takeWhile((res) => res.isFetching, true);
}
/**
 * An operator that takes values emitted by the source observable
 * until the `isSuccess` property on the result is true.
 * It is intended to be used in scenarios where an observable stream should be listened to
 * until a successful result is emitted.
 */
export function takeUntilResultSuccess() {
    return takeWhile((res) => !res.isSuccess, true);
}
/**
 * An operator that takes values emitted by the source observable
 * until the `isError` property on the result is true.
 * It is intended to be used in scenarios where an observable stream should be listened to
 * until an error result is emitted.
 */
export function takeUntilResultError() {
    return takeWhile((res) => !res.isError, true);
}
export function startWithPendingQueryResult() {
    return startWith(createPendingObserverResult());
}
/**
 *
 *  This operator is used to merge multiple queries into one.
 *  It will return a new base query result that will merge the results of all the queries.
 *
 * @example
 *
 * const query = combineLatest({
 *   todos: todos.result$,
 *   posts: posts.result$,
 * }).pipe(
 *   intersectResults$(({ todos, posts }) => {
 *     return { ... }
 *   })
 * )
 * @example
 *
 * const query = combineLatest([todos.result$, posts.result$]).pipe(
 *   intersectResults$(([todos, posts]) => {
 *     return { ... }
 *   })
 * )
 */
export function intersectResults$(mapFn) {
    return map((values) => {
        const isArray = Array.isArray(values);
        const toArray = isArray ? values : Object.values(values);
        const refetch = () => Promise.all(toArray.map(v => v.refetch()));
        const mappedResult = {
            all: values,
            isSuccess: toArray.every((v) => v.isSuccess),
            isPending: toArray.some((v) => v.isPending),
            isLoading: toArray.some((v) => v.isLoading),
            isError: toArray.some((v) => v.isError),
            isFetching: toArray.some((v) => v.isFetching),
            error: toArray.find((v) => v.isError)?.error,
            data: undefined,
            refetch,
        };
        if (mappedResult.isSuccess) {
            if (isArray) {
                mappedResult.data = mapFn(toArray.map((r) => r.data));
            }
            else {
                const data = Object.entries(values).reduce((acc, [key, value]) => {
                    acc[key] = value.data;
                    return acc;
                }, {});
                mappedResult.data = mapFn(data);
            }
        }
        return mappedResult;
    });
}
