// @ngneat/query/esm2022/lib/query.mjs
import { assertInInjectionContext, inject, Injectable, Injector, runInInjectionContext, } from '@angular/core';
import { injectQueryClient } from './query-client';
import { QueryObserver, } from '@tanstack/query-core';
import { createBaseQuery } from './base-query';
import * as i0 from "@angular/core";
class Query {
    #instance = injectQueryClient();
    #injector = inject(Injector);
    use(options) {
        return createBaseQuery({
            client: this.#instance,
            injector: options.injector ?? this.#injector,
            Observer: QueryObserver,
            options,
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.1.3", ngImport: i0, type: Query, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.1.3", ngImport: i0, type: Query, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.1.3", ngImport: i0, type: Query, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }] });
/**
 *
 * Optionally pass an injector that will be used than the current one.
 * Can be useful if you want to use it in ngOnInit hook for example.
 *
 * @example
 *
 * injector = inject(Injector);
 *
 * ngOnInit() {
 *  const todos = getTodos({ injector: this.injector }).result;
 * }
 *
 */
export function injectQuery(options) {
    if (options?.injector) {
        return runInInjectionContext(options.injector, () => {
            const query = inject(Query);
            return query.use.bind(query);
        });
    }
    assertInInjectionContext(injectQuery);
    const query = inject(Query);
    return query.use.bind(query);
}
