// @ngneat/query/esm2022/lib/signals.mjs
import { computed } from '@angular/core';
/**
 *
 *  @experimental
 *
 *  This function is used to merge multiple signal queries into one.
 *  It will return a new base query result that will merge the results of all the queries.
 *  Note that it should be used inside injection context
 *
 * @example
 *
 * const query = intersetResults({
 *   todos: todos.result$,
 *   posts: posts.result$,
 * }, ({ todos, posts }) => {
 *   return todos + posts;
 * })
 *
 *
 * @example
 *
 * const query = intersectResults(
 *   [
 *     this.todosService.getTodo('1').result,
 *     this.todosService.getTodo('2').result,
 *   ],
 *  ([todoOne, todoTwo]) => {
 *    return todoOne.title + todoTwo.title;
 *  }
 * );
 */
export function intersectResults(signals, mapFn) {
    const isArray = Array.isArray(signals);
    const toArray = isArray ? signals : Object.values(signals);
    const refetch = () => Promise.all(toArray.map(v => v().refetch()));
    return computed(() => {
        const mappedResult = {
            all: signals,
            isSuccess: toArray.every((v) => v().isSuccess),
            isPending: toArray.some((v) => v().isPending),
            isLoading: toArray.some((v) => v().isLoading),
            isError: toArray.some((v) => v().isError),
            isFetching: toArray.some((v) => v().isFetching),
            error: toArray.find((v) => v().isError)?.error,
            data: undefined,
            refetch,
        };
        if (mappedResult.isSuccess) {
            if (isArray) {
                mappedResult.data = mapFn(toArray.map((r) => r().data));
            }
            else {
                const data = Object.entries(signals).reduce((acc, [key, value]) => {
                    acc[key] = value().data;
                    return acc;
                }, {});
                mappedResult.data = mapFn(data);
            }
        }
        return mappedResult;
    });
}
