// @ngneat/query/esm2022/lib/utils.mjs
import { Subject, firstValueFrom, takeUntil } from 'rxjs';
export function toPromise({ source, signal, }) {
    const cancel = new Subject();
    if (signal) {
        signal.addEventListener('abort', () => {
            cancel.next();
            cancel.complete();
        });
    }
    return firstValueFrom(source.pipe(signal ? takeUntil(cancel) : (s) => s));
}
export function createSuccessObserverResult(data) {
    return {
        data,
        isLoading: false,
        isError: false,
        isFetching: false,
        isPending: false,
        isSuccess: true,
        status: 'success',
    };
}
export function createPendingObserverResult() {
    return {
        isError: false,
        isLoading: true,
        isPending: true,
        isFetching: true,
        isSuccess: false,
        fetchStatus: 'fetching',
        status: 'pending',
    };
}
export function shouldThrowError(throwError, params) {
    if (typeof throwError === 'function') {
        return throwError(...params);
    }
    return !!throwError;
}
